### PR TITLE
Support user-provided timeouts for request-response messages

### LINKED ISSUE #477: Is there a way to set timeout for ClusterMessagingService.send method?
Could you explain,  Is there a way to set timeout for responce for ClusterMessagingService.send method? and what is the default timeout?
I haven't found documentation about this.

### PR DESCRIPTION
**Pull request by kuujo**
This PR fixes #477. We just implement `Duration` based timeouts in `NettyMessagingService` and expose the timeout to all higher level point-to-point messaging methods. If `Duration` is `null` then the dynamic timeout is used.

### CODE REVIEW

File: core/src/test/java/io/atomix/core/TestMessagingService.java [hunk 1]
```diff
@@ -76,19 +77,31 @@ public Address address() {
...
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Duration timeout) {
+    return getHandler(address, type).apply(address, payload);
   }
 
   @Override
-  public CompletableFuture<byte[]> sendAndReceive(Address ep, String type, byte[] payload, Executor executor) {
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Duration timeout, Executor executor) {
     ComposableFuture<byte[]> future = new ComposableFuture<>();
-    sendAndReceive(ep, type, payload).whenCompleteAsync(future, executor);
+    sendAndReceive(address, type, payload).whenCompleteAsync(future, executor);
```
Comment (johnou) [inline core/src/test/java/io/atomix/core/TestMessagingService.java]: Shouldn't we be passing timeout into send and receive here?

File: cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java [hunk 2]
```diff
@@ -76,19 +77,31 @@ public Address address() {
...
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Executor executor) {
     ComposableFuture<byte[]> future = new ComposableFuture<>();
-    sendAndReceive(ep, type, payload).whenCompleteAsync(future, executor);
+    sendAndReceive(address, type, payload).whenCompleteAsync(future, executor);
+    return future;
+  }
+
+  @Override
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Duration timeout) {
+    return getHandler(address, type).apply(address, payload);
```
Comment (johnou) [inline cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java]: Ignored timeout

File: cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java [hunk 3]
```diff
@@ -76,19 +77,31 @@ public Address address() {
...
+  }
+
+  @Override
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Duration timeout, Executor executor) {
+    ComposableFuture<byte[]> future = new ComposableFuture<>();
+    sendAndReceive(address, type, payload).whenCompleteAsync(future, executor);
```
Comment (johnou) [inline cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java]: Ignored timeout

File: core/src/test/java/io/atomix/core/TestMessagingService.java [hunk 4]
```diff
@@ -76,19 +77,31 @@ public Address address() {
...
+    ComposableFuture<byte[]> future = new ComposableFuture<>();
+    sendAndReceive(address, type, payload).whenCompleteAsync(future, executor);
+    return future;
   }
 
   @Override
-  public CompletableFuture<byte[]> sendAndReceive(Address ep, String type, byte[] payload) {
-    return getHandler(ep, type).apply(address, payload);
+  public CompletableFuture<byte[]> sendAndReceive(Address address, String type, byte[] payload, Duration timeout) {
+    return getHandler(address, type).apply(address, payload);
```
Comment (johnou) [inline core/src/test/java/io/atomix/core/TestMessagingService.java]: Ignored timeout

File: core/src/test/java/io/atomix/core/TestMessagingService.java (hunk 1 above)
Comment (kuujo) [reply core/src/test/java/io/atomix/core/TestMessagingService.java]: The `TestMessagingService` doesn't time out any messages ATM

File: cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java (hunk 2 above)
Comment (kuujo) [reply cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java]: Same

File: core/src/test/java/io/atomix/core/TestMessagingService.java (hunk 4 above)
Comment (kuujo) [reply core/src/test/java/io/atomix/core/TestMessagingService.java]: Same

File: cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java (hunk 3 above)
Comment (kuujo) [reply cluster/src/test/java/io/atomix/cluster/messaging/impl/TestMessagingService.java]: Same

**Comment by kuujo** [comment]
I removed one `TestMessagingService`, and the other isn't used for tests that rely on timeouts.

**Comment by kuujo** [comment]
One of the tests seems to be consistently failing with this change. Can't tell why

**Comment by johnou** [comment]
dataClusterService.getNodes() is no longer returning the client node.

**Comment by kuujo** [comment]
Turns out it was unrelated, but fixed.

File: cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterService.java [hunk 5]
```diff
@@ -183,7 +183,7 @@ private void sendHeartbeats() {
       sendHeartbeat(node.address(), payload);
       PhiAccrualFailureDetector failureDetector = failureDetectors.computeIfAbsent(node.id(), n -> new PhiAccrualFailureDetector());
       double phi = failureDetector.phi();
-      if (phi >= phiFailureThreshold || System.currentTimeMillis() - failureDetector.lastUpdated() > DEFAULT_FAILURE_TIME) {
+      if (phi > 0 && (phi >= phiFailureThreshold || System.currentTimeMillis() - failureDetector.lastUpdated() > DEFAULT_FAILURE_TIME)) {
```
Comment (johnou) [inline cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterService.java]: @kuujo wouldn't 100 make sense here to keep it consistent with the Netty failure detection minimum value?

Comment (johnou) [reply cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterService.java]: oops no, phi, not timeout ms.

**Pull request merged**